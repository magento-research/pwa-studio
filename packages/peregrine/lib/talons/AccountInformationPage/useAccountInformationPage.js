import { useCallback, useMemo, useState } from 'react';
import { useMutation, useQuery } from '@apollo/client';

import { useAppContext } from '../../context/app';
import { useUserContext } from '../../context/user';

export const useAccountInformationPage = props => {
    const {
        // afterSubmit,
        // mutations: { setNewsletterSubscriptionMutation },
        queries: { getCustomerInformationQuery }
    } = props;

    const [, { toggleDrawer }] = useAppContext();
    const [{ isSignedIn }] = useUserContext();

    const [activeInformation, setActiveInformation] = useState();

    const { data: accountInformationData, error: accountInformationError } = useQuery(
        getCustomerInformationQuery,
        { skip: !isSignedIn, fetchPolicy: 'cache-and-network' }
    );

    const handleEditInformation = useCallback(
        info => {
            setActiveInformation(info);
            toggleDrawer('accountInformation.edit');
        },
        [toggleDrawer]
    );

    const initialValues = useMemo(() => {
        if (accountInformationData) {
            return { customer: accountInformationData.customer };
        }
    }, [accountInformationData]);

    return {
        initialValues,
        formErrors: [accountInformationError],
        isSignedIn,
        handleEditInformation
    }
}
