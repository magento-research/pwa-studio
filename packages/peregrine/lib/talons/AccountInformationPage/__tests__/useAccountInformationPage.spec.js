import React from 'react';
import { useQuery, useMutation } from '@apollo/client';

import createTestInstance from '../../../util/createTestInstance';
import { useAccountInformationPage } from '../useAccountInformationPage';

const mockSetCustomerInformation = jest.fn();
const mockChangeCustomerPassword = jest.fn();

jest.mock('@magento/peregrine/lib/context/user', () => {
    const state = {
        isSignedIn: true
    };
    const api = {};
    const useUserContext = jest.fn(() => [state, api]);

    return { useUserContext };
});

jest.mock('@apollo/client', () => ({
    useMutation: jest.fn().mockImplementation(mutation => {
        if (mutation === 'setCustomerInformationMutation')
            return [mockSetCustomerInformation, { loading: false }];

        if (mutation === 'changeCustomerPasswordMutation')
            return [mockChangeCustomerPassword, { loading: false }];

        return;
    }),
    useQuery: jest.fn().mockReturnValue({
        data: {
            customer: {
                id: null,
                firtname: 'Foo',
                lastname: 'Bar',
                email: 'foobar@express.net'
            }
        },
        error: null,
        loading: false
    })
}));

const Component = props => {
    const talonProps = useAccountInformationPage(props);
    return <i talonProps={talonProps} />;
};

const mockProps = {
    mutations: {
        setCustomerInformationMutation: 'setCustomerInformationMutation',
        changeCustomerPasswordMutation: 'changeCustomerPasswordMutation'
    },
    queries: {
        getCustomerInformationQuery: 'getCustomerInformationQuery'
    }
};

test('return correct shape while data is loading', () => {
    useQuery.mockReturnValueOnce({
        loading: true
    });

    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('return correct shape for initial customer data', () => {
    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('handleChangePassword sets shouldShowNewPassword to true', () => {});

test('showUpdateMode sets isUpdateMode to true', () => {});

test('handleSubmit calls setCustomerInformationQuery', async () => {
    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;
    const { handleSubmit } = talonProps;

    await handleSubmit({
        firtname: 'Foo',
        lastname: 'Bar',
        email: 'foobar@express.net'
    });

    expect(mockSetCustomerInformation).toHaveBeenCalled();
    expect(mockChangeCustomerPassword).not.toHaveBeenCalled();
    expect(mockSetCustomerInformation.mock.calls[0][0]).toMatchSnapshot();
});

test('handleSubmit calls changeCustomerPassword if new password is provided', async () => {
    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;
    const { handleSubmit } = talonProps;

    await handleSubmit({
        firtname: 'Foo',
        lastname: 'Bar',
        email: 'foobar@express.net',
        password: 'bar',
        newPassword: 'notBar'
    });

    expect(mockChangeCustomerPassword).toHaveBeenCalled();
    expect(mockChangeCustomerPassword.mock.calls[0][0]).toMatchSnapshot();
});

test('handleSubmit does not throw', async () => {
    const mockSetCustomerInformation = jest
        .fn()
        .mockRejectedValue(new Error('Async error'));
    useMutation.mockReturnValue([mockSetCustomerInformation, {}]);

    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;
    const { handleSubmit } = talonProps;

    expect(async () => {
        await handleSubmit({
            firtname: 'Foo',
            lastname: 'Bar',
            email: 'foobar@express.net'
        });
    }).not.toThrow();

    expect(mockSetCustomerInformation).toHaveBeenCalled();
    expect(mockChangeCustomerPassword).not.toHaveBeenCalled();
});
