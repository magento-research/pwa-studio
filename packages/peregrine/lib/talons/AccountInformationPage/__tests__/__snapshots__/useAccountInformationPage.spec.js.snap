// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return correct shape for initial customer data 1`] = `
Object {
  "cancelUpdateMode": [Function],
  "formErrors": Array [
    undefined,
    undefined,
  ],
  "handleChangePassword": [Function],
  "handleSubmit": [Function],
  "initialValues": Object {
    "customer": Object {
      "email": "foobar@express.net",
      "firtname": "Foo",
      "id": null,
      "lastname": "Bar",
    },
  },
  "isDisabled": false,
  "isSignedIn": true,
  "isUpdateMode": false,
  "loadDataError": null,
  "shouldShowNewPassword": false,
  "showUpdateMode": [Function],
}
`;

exports[`return correct shape for new value and fire create mutation update customer information 1`] = `
Object {
  "cancelUpdateMode": [Function],
  "formErrors": Array [
    undefined,
    undefined,
  ],
  "handleChangePassword": [Function],
  "handleSubmit": [Function],
  "initialValues": Object {
    "customer": Object {
      "email": "barfoo@express.net",
      "firtname": "Bar",
      "id": 7,
      "lastname": "Foo",
    },
  },
  "isDisabled": false,
  "isSignedIn": true,
  "isUpdateMode": false,
  "loadDataError": null,
  "shouldShowNewPassword": false,
  "showUpdateMode": [Function],
}
`;

exports[`return correct shape for new value and fire create mutation update customer information 2`] = `
Object {
  "variables": Object {
    "email": "foobar@express.net",
    "firtname": "Foo",
    "lastname": "Bar",
  },
}
`;

exports[`return correct shape for new value and fire create mutation update customer information and update new password 1`] = `
Object {
  "cancelUpdateMode": [Function],
  "formErrors": Array [
    undefined,
    undefined,
  ],
  "handleChangePassword": [Function],
  "handleSubmit": [Function],
  "initialValues": Object {
    "customer": Object {
      "email": "barfoo@express.net",
      "firtname": "Bar",
      "id": 7,
      "lastname": "Foo",
    },
  },
  "isDisabled": false,
  "isSignedIn": true,
  "isUpdateMode": false,
  "loadDataError": null,
  "shouldShowNewPassword": false,
  "showUpdateMode": [Function],
}
`;

exports[`return correct shape for new value and fire create mutation update customer information and update new password 2`] = `
Object {
  "variables": Object {
    "email": "foobar@express.net",
    "firtname": "Foo",
    "lastname": "Bar",
    "newPassword": "Gooseton123@",
    "password": "Gooseton123?",
  },
}
`;

exports[`return correct shape for new value and fire create mutation update customer information and update new password 3`] = `
Object {
  "variables": Object {
    "currentPassword": "Gooseton123?",
    "newPassword": "Gooseton123@",
  },
}
`;

exports[`return correct shape while data is loading 1`] = `
Object {
  "cancelUpdateMode": [Function],
  "formErrors": Array [
    undefined,
    undefined,
  ],
  "handleChangePassword": [Function],
  "handleSubmit": [Function],
  "initialValues": undefined,
  "isDisabled": false,
  "isSignedIn": true,
  "isUpdateMode": false,
  "loadDataError": undefined,
  "shouldShowNewPassword": false,
  "showUpdateMode": [Function],
}
`;
